### PR TITLE
compute lengths cache in concat

### PR DESCRIPTION
**Pull request by anmyachev**
With changes in https://github.com/modin-project/modin/pull/4816 we can make the execution of read_parquet + concat asynchronous. It will be useful when reading several parquet files.

Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [ ] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [ ] passes `flake8 modin`
- [ ] passes `black --check modin`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #? 
- [ ] tests added and passing

### CODE REVIEW

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 1]
```diff
@@ -1334,17 +1349,12 @@ def rebalance_partitions(cls, partitions):
...
             num_existing_partitions = partitions.shape[0]
             ideal_num_new_partitions = NPartitions.get()
-            if (
-                num_existing_partitions
-                <= ideal_num_new_partitions * max_excess_of_num_partitions
-            ):
+            need_rebalance = cls._need_rebalance(num_existing_partitions)
+            if not need_rebalance:
                 return partitions
+            max_excess_of_num_partitions, ideal_num_new_partitions = need_rebalance
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1357]: need_rebalance is a boolean object. What do we do here?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 2]
```diff
@@ -1308,6 +1308,21 @@ def finalize(cls, partitions):
         """
         [part.drain_call_queue() for row in partitions for part in row]
 
+    @classmethod
+    def _need_rebalance(cls, count_row_partitions):
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1312]:
```suggestion
    def _need_rebalance(cls, row_partitions_count):
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 3]
```diff
@@ -1308,6 +1308,21 @@ def finalize(cls, partitions):
...
+        # We rebalance when the ratio of the number of existing partitions to
+        # the ideal number of partitions is larger than this threshold. The
+        # threshold is a heuristic that may need to be tuned for performance.
+        max_excess_of_num_partitions = 1.5
+        num_existing_partitions = count_row_partitions
+        ideal_num_new_partitions = NPartitions.get()
+        if (
+            num_existing_partitions
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1320]:
```suggestion
        ideal_num_new_partitions = NPartitions.get()
        if (
            row_partitions_count
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 4]
```diff
@@ -1334,17 +1349,12 @@ def rebalance_partitions(cls, partitions):
...
-            # the ideal number of partitions is larger than this threshold. The
-            # threshold is a heuristic that may need to be tuned for performance.
-            max_excess_of_num_partitions = 1.5
             num_existing_partitions = partitions.shape[0]
             ideal_num_new_partitions = NPartitions.get()
-            if (
-                num_existing_partitions
-                <= ideal_num_new_partitions * max_excess_of_num_partitions
-            ):
+            need_rebalance = cls._need_rebalance(num_existing_partitions)
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:1354]:
```suggestion
            need_rebalance = cls._needs_rebalance(num_existing_partitions)
```

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 5]
```diff
@@ -2652,6 +2653,24 @@ def _compute_new_widths():
...
+            row_partitions_count = sum(
+                [
+                    part.shape[0]
+                    for part in [self._partitions] + [o._partitions for o in others]
+                ]
+            )
+            if (
+                self._row_lengths_cache is not None
+                and all(o._row_lengths_cache is not None for o in others)
+                and not self._partition_mgr_cls._need_rebalance(row_partitions_count)
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:2667]:
```suggestion
                and not self._partition_mgr_cls._needs_rebalance(row_partitions_count)
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 1 above)
Comment (anmyachev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1356]: _needs_rebalance returns "False" or data needed for rebalance.

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:1356]: I see, thanks!

**Pull request merged**